File: week02/이분탐색/2805.py
# N : 나무 개수 (N <= 1,000,000)
# M : 집으로 가져가야하는 나무의 길이
# H : 톱날의 높이 (H <= 2,000,000,000)
# 구해야 할 것 : M미터의 나무를 집에 가져가기 위한 H의 최댓값
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
tree_list = list(map(int, input().split())) 

# tree_list.sort() 
start = 0 # 4
end = max(tree_list)

# 첫 임시 톱날 높이(temp_h) => (start + end) // 2
# temp_h로 잘랐을때 나오는 나무 길이인 temp_M.
# temp_M > M 일경우, 톱날의 높이는 더 높아야함, 따라서 temp_H ~ end까지
# temp_M < M 일경우 톱날의 높이는 더 낮아야함, 따라서 low ~ temp_H까지

def binary_search(data, start, end):
  temp_h = (start + end) // 2 # 4 + 46 // 2 => 25
  temp_M = 0
  if start > end: # 나중에 34 35일 경우 종료조건으로 진행
    return end
  
  for tree in data:
    if temp_h > tree:
      continue
    else:
      temp_M += tree - temp_h
    
  if temp_M == M:
    return temp_h
  elif temp_M > M:
    return binary_search(data, temp_h + 1, end)
  else:
    return binary_search(data, start, temp_h -1)
  

print(binary_search(tree_list, start, end))

File: week02_JS/이분탐색/2805.js
// 나무 높이를 정렬
// 나무 높이의 중앙값 설정
// 그 높이로 자른 값과 가져가야할 높이 H를 비교
// 그 높이로 자른 값이 더 크면 => 높이가 더 높아야함
// 그 높이로 자른 값이 더 작다면 => 높이가 더 낮아야함
const fs = require('fs')
const input = fs.readFileSync('input.txt').toString().trim().split('\n');

let [N, M] = input[0].split(' ').map(Number)
let tree_list = input[1].split(' ').map(Number)

tree_list.sort((a, b) => a - b)

function binary_search(arr, M) {
  let start = 0
  let end = arr[arr.length - 1]  // 나무의 최대 높이로 설정

  while (start <= end) {
    let mid = Math.floor((start + end) / 2)

    // mid 값으로 자른 나무의 길이 합을 계산
    let temp = arr.reduce((sum, height) => {
      if (height > mid) return sum + (height - mid)
      return sum
    }, 0)

    if (temp === M) {
      console.log(mid)
      return
    } else if (temp > M) {
      start = mid + 1  // 너무 많이 잘랐으니, 높이를 더 높여야 함
    } else {
      end = mid - 1    // 너무 적게 잘랐으니, 높이를 더 낮춰야 함
    }
  }

  console.log(end)
}

binary_search(tree_list, M)
